app/components/search.tsx

```diff
@@ -23,12 +23,11 @@ type SearchResult = Question & {
   url?: string
 }
 
-type WorkerMessage =
```

Comment:
[NIT] it seems a bit iffy to have this as two separate types

Reply:
Good point! I was so fixated on getting rid of the TypeScript errors that I didn't think through the fundamentals. Well, I tried this...
```
type WorkerMessage =
  | { status: string; numQs: number }
  | { searchResults: {title: string; pageid: string; score: number}[]; userQuery: string }
```
but it's still giving me these errors:
```
app/components/search.tsx:45:16 - error TS2339: Property 'status' does not exist on type 'WorkerMessage'.
  Property 'status' does not exist on type '{ searchResults: { title: string; pageid: string; score: number; }[]; userQuery: string; }'.
45       if (data.status === 'ready') {
                  ~~~~~~
app/components/search.tsx:49:16 - error TS2339: Property 'searchResults' does not exist on type 'WorkerMessage'.
  Property 'searchResults' does not exist on type '{ status: string; numQs: number; }'.
49       if (data.searchResults) {
                  ~~~~~~~~~~~~~
app/components/search.tsx:50:31 - error TS2339: Property 'searchResults' does not exist on type 'WorkerMessage'.
  Property 'searchResults' does not exist on type '{ status: string; numQs: number; }'.
50         setSearchResults(data.searchResults)
```
So I'm obviously missing something. To be honest, my understanding of TypeScript is purely by trial and error. How would you suggest to fix this properly?

Reply:
It think something like this should work:
```
type WorkerMessage = {
  searchResults?: {title: string; pageid: string; score: number}[]
  status: string
  numQs?: number
}
```

Reply:
Ah, that did it, thanks! Aprillion tried explaining to me what the ? meant but I guess it didn't stick.

Reply:
yup, typescript is annoying in that way :D

Reply:
typescript is nice, when all types are shared ... but the worker is native JavaScript file (executed as-is without compilation in the browser for now), so the TS type must match what the worker actually sends and there is no TS error if we forget to synchronize both

feel free to refactor to compile the workef rom TS you want to make it less annoying 😸

Reply:
actually, the type doesn't match - following lines post message from the worker:
```js
23: self.postMessage({status: 'ready', numQs: questions.length})
39: self.postMessage({searchResults: []})
68: self.postMessage({searchResults, userQuery})
```

the correct type for that is:
```ts
type WorkerMessage = {
  status: 'ready'
  numQs: number
} | {
  searchResults: {title: string, pageid: string, score: number}[]
  userQuery?: string
}
```

and fixing the type will probably produce TS errors in this file => @ccstan99 let me know if you want me to fix all TS thingies myself

Reply:
Yes, pretty please! I'd love to learn TS (and a bunch of other things) properly some day. I can definitely see its potential for beauty and power. Unfortunately, I don't have the bandwidth today.

Reply:
@Aprillion Should I wait for you to take care of the TS thingies before merging? Or should we merge first and cleanup later? Otherwise, I'll need to find some time to go back to my trial and error approach with TS. Would you mind sending me that link that explains how those ? work in TS again? I forgot what they're even called. I remember it made sense initially, but unfortunately, old age & memory don't go well together.